src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs

```diff
@@ -0,0 +1,150 @@
+
+                return reference.Value;
+            }
+            else
+            {
+                foreach (var member in _members)
+                    await member.Read(composite, buffer, async);
+
+                return composite;
+            }
+        }
+
+        public override async Task Write(T value, NpgsqlWriteBuffer buffer, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        {
+            if (buffer.WriteSpaceLeft < sizeof(int))
+                await buffer.Flush(async);
+
+            buffer.WriteInt32(_members.Length);
+
+            foreach (var member in _members)
+                await member.Write(value, buffer, lengthCache, async);
+        }
+
+        public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+        {
+            if (lengthCache == null)
+                lengthCache = new NpgsqlLengthCache(1);
+
+            if (lengthCache.IsPopulated)
+                return lengthCache.Get();
+
+            // Leave empty slot for the entire composite type, and go ahead an populate the element slots
+            var position = lengthCache.Position;
+            lengthCache.Set(0);
+
+            // number of fields + (type oid + field length) * member count
+            var length = sizeof(int) + sizeof(int) * 2 * _members.Length;
+            foreach (var member in _members)
+                length += member.ValidateAndGetLength(value, ref lengthCache);
+
+            return lengthCache.Lengths[position] = length;
+        }
+
+        public static MappedCompositeHandler<T> Create(PostgresCompositeType pgType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        {
+            var clrType = typeof(T);
+            var pgFields = pgType.Fields;
+            var clrHandlers = new CompositeMemberHandler<T>[pgFields.Count];
+            var clrHandlerCount = 0;
+
+            var clrHandlerType = IsValueType<T>.Value
+                ? typeof(CompositeMemberHandlerOfStruct<,>)
+                : typeof(CompositeMemberHandlerOfClass<,>);
+
+            foreach (var clrMember in clrType.GetMembers(BindingFlags.Instance | BindingFlags.Public))
+            {
+                Type clrMemberType;
+                switch (clrMember)
+                {
+                    case FieldInfo clrField:
+                        clrMemberType = clrField.FieldType;
+                        break;
+                    case PropertyInfo clrProperty:
+                        clrMemberType = clrProperty.PropertyType;
+                        break;
+                    default:
+                        continue;
+                }
+                
+                var attr = clrMember.GetCustomAttribute<PgNameAttribute>();
+                var name = attr != null
+                    ? attr.PgName
+                    : nameTranslator.TranslateMemberName(clrMember.Name);
```

Comment:
Unfortunately, these expressions aren't equal to each other. `PgNameAttribute` doesn't check that its argument. Therefore, the name can be `null`. I think in that case we should fail with a field not mapped exception.

Probably the behavior should be changed, but let's do it as a separate pull request. Otherwise, there is a change that I will refactor the whole driver here (: Don't resolve this comment, I'll open an issue to track that.

Reply:
> Unfortunately, these expressions aren't equal to each other.

You are correct, but I would still opt to treat them the same anyways. For one thing, it's [what we do in the EF Core provider](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/6b58e6c280d15318aebbd8e259bb7c532cccfdd1/src/EFCore.PG/Storage/Internal/Mapping/NpgsqlEnumTypeMapping.cs#L61).

> I think in that case we should fail with a field not mapped exception.

That seems a little too aggressive to me. If it's a typo (e.g. forgot to add the name) and the CLR name isn't going to map, we'll fail soon anyways. So why not allow for some flexibility by falling back to the name translator at this point?

Reply:
Makes sense, will change.